File: app/src/main/java/org/gnucash/android/util/BackupManager.java
/* Copyright (c) 2018 Àlex Magaz Graça <alexandre.magaz@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gnucash.android.util;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.AsyncTask;
import android.text.TextUtils;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;
import androidx.work.WorkRequest;

import org.gnucash.android.R;
import org.gnucash.android.app.GnuCashApplication;
import org.gnucash.android.db.adapter.BooksDbAdapter;
import org.gnucash.android.export.ExportFormat;
import org.gnucash.android.export.ExportParams;
import org.gnucash.android.export.Exporter;
import org.gnucash.android.export.xml.GncXmlExporter;
import org.gnucash.android.model.Book;
import org.gnucash.android.ui.common.GnucashProgressDialog;
import org.gnucash.android.ui.settings.PreferenceActivity;
import org.gnucash.android.work.BackupWorker;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.zip.GZIPOutputStream;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import timber.log.Timber;

/**
 * Deals with all backup-related tasks.
 */
public class BackupManager {

    public static final String KEY_BACKUP_FILE = "book_backup_file_key";
    public static final String MIME_TYPE = "application/gzip";

    /**
     * Perform an automatic backup of all books in the database.
     * This method is run every time the service is executed
     */
    @WorkerThread
    static boolean backupAllBooks() {
        Context context = GnuCashApplication.getAppContext();
        return backupAllBooks(context);
    }

    /**
     * Perform an automatic backup of all books in the database.
     * This method is run every time the service is executed
     * @return `true` when all books were successfully backed-up.
     */
    @WorkerThread
    public static boolean backupAllBooks(Context context) {
        Timber.i("Doing backup of all books.");
        BooksDbAdapter booksDbAdapter = BooksDbAdapter.getInstance();
        List<String> bookUIDs = booksDbAdapter.getAllBookUIDs();

        for (String bookUID : bookUIDs) {
            if (!backupBook(context, bookUID)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Backs up the active book to the directory {@link #getBackupFolder(String)}.
     *
     * @return {@code true} if backup was successful, {@code false} otherwise
     */
    @WorkerThread
    public static boolean backupActiveBook() {
        return backupActiveBook(GnuCashApplication.getAppContext());
    }

    /**
     * Backs up the active book to the directory {@link #getBackupFolder(String)}.
     *
     * @return {@code true} if backup was successful, {@code false} otherwise
     */
    @WorkerThread
    public static boolean backupActiveBook(Context context) {
        return backupBook(context, GnuCashApplication.getActiveBookUID());
    }

    /**
     * Backs up the book with UID {@code bookUID} to the directory
     * {@link #getBackupFolder(String)}.
     *
     * @param bookUID Unique ID of the book
     * @return {@code true} if backup was successful, {@code false} otherwise
     */
    @WorkerThread
    public static boolean backupBook(Context context, String bookUID) {
        ExportParams params = new ExportParams(ExportFormat.XML);
        OutputStream outputStream;
        try {
            Uri backupUri = getBookBackupFileUri(bookUID);
            if (backupUri != null) {
                outputStream = context.getContentResolver().openOutputStream(backupUri);
            } else { //no Uri set by user, use default location on SD card
                File backupFile = getBackupFile(bookUID, params);
                outputStream = new FileOutputStream(backupFile);
            }
            params.setExportLocation(backupUri);

            BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(outputStream);
            GZIPOutputStream gzipOutputStream = new GZIPOutputStream(bufferedOutputStream);
            OutputStreamWriter writer = new OutputStreamWriter(gzipOutputStream);

            new GncXmlExporter(context, params, bookUID).generateExport(writer);
            writer.close();
            return true;
        } catch (Throwable e) {
            Timber.e(e, "Error creating backup");
        }
        return false;
    }

    /**
     * Returns the full path of a file to make database backup of the specified book.
     * Backups are done in XML format and are Gzipped (with ".gnucash" extension).
     *
     * @param bookUID GUID of the book
     * @param params the export parameters.
     * @return the file for backups of the database.
     * @see #getBackupFolder(String)
     */
    private static File getBackupFile(@NonNull String bookUID, @Nullable ExportParams params) {
        Book book = BooksDbAdapter.getInstance().getRecord(bookUID);
        ExportFormat format = (params != null) ? params.getExportFormat() : ExportFormat.XML;
        String name = Exporter.buildExportFilename(format, book.getDisplayName()) + ".gz";
        return new File(getBackupFolder(book.getUID()), name);
    }

    /**
     * Returns the path to the backups folder for the book with GUID {@code bookUID}.
     *
     * <p>Each book has its own backup folder.</p>
     *
     * @return The backup folder for the book
     */
    private static File getBackupFolder(String bookUID) {
        File baseFolder = GnuCashApplication.getAppContext().getExternalFilesDir(null);
        File folder = new File(baseFolder, bookUID + File.separator + "backups");
        if (!folder.exists())
            folder.mkdirs();
        return folder;
    }

    /**
     * Return the user-set backup file URI for the book with UID {@code bookUID}.
     *
     * @param bookUID Unique ID of the book
     * @return DocumentFile for book backups, or null if the user hasn't set any.
     */
    @Nullable
    public static Uri getBookBackupFileUri(String bookUID) {
        SharedPreferences sharedPreferences = PreferenceActivity.getBookSharedPreferences(bookUID);
        String path = sharedPreferences.getString(KEY_BACKUP_FILE, null);
        if (TextUtils.isEmpty(path)) {
            return null;
        }
        return Uri.parse(path);
    }

    public static List<File> getBackupList(String bookUID) {
        File[] backupFiles = getBackupFolder(bookUID).listFiles();
        Arrays.sort(backupFiles);
        List<File> backupFilesList = Arrays.asList(backupFiles);
        Collections.reverse(backupFilesList);
        return backupFilesList;
    }

    public static void schedulePeriodicBackups(Context context) {
        Timber.i("Scheduling backups");
        WorkRequest request = new PeriodicWorkRequest.Builder(BackupWorker.class, 1, TimeUnit.DAYS)
            .setInitialDelay(15, TimeUnit.MINUTES)
            .build();

        WorkManager.getInstance(context)
            .enqueue(request);
    }

    public static void backupBookAsync(@Nullable final Activity activity, final String bookUID, @NonNull final Function1<Boolean, Unit> after) {
        new AsyncTask<Object, Void, Boolean>() {
            private ProgressDialog mProgressDialog;

            @Override
            protected void onPreExecute() {
                if (activity != null) {
                    mProgressDialog = new GnucashProgressDialog(activity);
                    mProgressDialog.setTitle(R.string.title_create_backup_pref);
                    mProgressDialog.setCancelable(true);
                    mProgressDialog.setOnCancelListener(dialogInterface -> cancel(true));
                    mProgressDialog.show();
                }
            }

            @Override
            protected Boolean doInBackground(Object... objects) {
                return backupBook(activity, bookUID);
            }

            @Override
            protected void onPostExecute(Boolean result) {
                try {
                    if (mProgressDialog != null && mProgressDialog.isShowing()) {
                        mProgressDialog.dismiss();
                    }
                } catch (IllegalArgumentException ex) {
                    //TODO: This is a hack to catch "View not attached to window" exceptions
                    //FIXME by moving the creation and display of the progress dialog to the Fragment
                } finally {
                    mProgressDialog = null;
                }
                if (!result) {
                    Toast.makeText(activity, R.string.toast_backup_failed, Toast.LENGTH_SHORT).show();
                }
                after.invoke(result);
            }

            @Override
            protected void onCancelled() {
                after.invoke(Boolean.FALSE);
            }
        }.execute();
    }

    public static void backupActiveBookAsync(@Nullable Activity activity, @NonNull final Function1<Boolean, Unit> after) {
        backupBookAsync(activity, GnuCashApplication.getActiveBookUID(), after);
    }
}

File: app/src/main/java/org/gnucash/android/util/BackupManager.kt
/* Copyright (c) 2018 Àlex Magaz Graça <alexandre.magaz@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.gnucash.android.util

import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.content.DialogInterface
import android.net.Uri
import android.os.AsyncTask
import androidx.annotation.WorkerThread
import androidx.core.net.toUri
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import androidx.work.WorkRequest
import org.gnucash.android.R
import org.gnucash.android.app.GnuCashApplication
import org.gnucash.android.db.adapter.BooksDbAdapter
import org.gnucash.android.export.ExportFormat
import org.gnucash.android.export.ExportParams
import org.gnucash.android.export.Exporter
import org.gnucash.android.export.xml.GncXmlExporter
import org.gnucash.android.lang.BooleanCallback
import org.gnucash.android.ui.common.GnucashProgressDialog
import org.gnucash.android.ui.snackLong
import org.gnucash.android.work.BackupWorker
import timber.log.Timber
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Deals with all backup-related tasks.
 */
object BackupManager {
    const val KEY_BACKUP_FILE: String = "book_backup_file_key"
    const val MIME_TYPE: String = "application/gzip"

    /**
     * Perform an automatic backup of all books in the database.
     * This method is run every time the service is executed
     */
    @WorkerThread
    fun backupAllBooks(): Boolean {
        val context = GnuCashApplication.appContext
        return backupAllBooks(context)
    }

    /**
     * Perform an automatic backup of all books in the database.
     * This method is run every time the service is executed
     *
     * @return `true` when all books were successfully backed-up.
     */
    @WorkerThread
    fun backupAllBooks(context: Context): Boolean {
        Timber.i("Doing backup of all books.")
        val booksDbAdapter = BooksDbAdapter.instance
        val bookUIDs = booksDbAdapter.allBookUIDs

        for (bookUID in bookUIDs) {
            if (!backupBook(context, bookUID)) {
                return false
            }
        }
        return true
    }

    /**
     * Backs up the active book to the directory [.getBackupFolder].
     *
     * @return `true` if backup was successful, `false` otherwise
     */
    @WorkerThread
    fun backupActiveBook(): Boolean {
        return backupActiveBook(GnuCashApplication.appContext)
    }

    /**
     * Backs up the active book to the directory [.getBackupFolder].
     *
     * @return `true` if backup was successful, `false` otherwise
     */
    @WorkerThread
    fun backupActiveBook(context: Context): Boolean {
        return backupBook(context, GnuCashApplication.activeBookUID!!)
    }

    /**
     * Backs up the book with UID `bookUID` to the directory
     * [.getBackupFolder].
     *
     * @param bookUID Unique ID of the book
     * @return `true` if backup was successful, `false` otherwise
     */
    @WorkerThread
    fun backupBook(context: Context, bookUID: String): Boolean {
        val params = ExportParams(ExportFormat.XML)
        params.exportTarget = ExportParams.ExportTarget.URI
        params.isCompressed = true
        try {
            val backupUri = getBookBackupFileUri(context, bookUID, params)
            params.exportLocation = backupUri
            val exporter: Exporter = GncXmlExporter(context, params, bookUID)
            val uri: Uri? = exporter.export()
            return (uri != null)
        } catch (e: Throwable) {
            Timber.e(e, "Error creating backup")
        }
        return false
    }

    /**
     * Returns the full path of a file to make database backup of the specified book.
     * Backups are done in XML format and are Gzipped (with ".gnucash" extension).
     *
     * @param context      the context
     * @param bookUID      GUID of the book
     * @param exportParams the export parameters.
     * @return the file for backups of the database.
     * @see .getBackupFolder
     */
    private fun getBackupFile(
        context: Context,
        bookUID: String,
        exportParams: ExportParams?
    ): File {
        val name = Exporter.buildExportFilename(exportParams, "backup")
        return File(getBackupFolder(context, bookUID), name)
    }

    /**
     * Returns the path to the backups folder for the book with GUID `bookUID`.
     *
     *
     * Each book has its own backup folder.
     *
     * @return The backup folder for the book
     */
    fun getBackupFolder(context: Context, bookUID: String): File {
        val baseFolder = context.getExternalFilesDir(null)
        val folder = File(File(baseFolder, bookUID), "backups")
        if (!folder.exists()) folder.mkdirs()
        return folder
    }

    /**
     * Return the user-set backup file URI for the book with UID `bookUID`.
     *
     * @param bookUID Unique ID of the book
     * @return DocumentFile for book backups, or null if the user hasn't set any.
     */
    fun getBookBackupFileUri(context: Context, bookUID: String): Uri? {
        return getBookBackupFileUri(context, bookUID, null)
    }

    /**
     * Return the user-set backup file URI for the book with UID `bookUID`.
     *
     * @param bookUID Unique ID of the book
     * @return DocumentFile for book backups, or null if the user hasn't set any.
     */
    fun getBookBackupFileUri(context: Context, bookUID: String, exportParams: ExportParams?): Uri? {
        val preferences = GnuCashApplication.getBookPreferences(context, bookUID)
        val path = preferences.getString(KEY_BACKUP_FILE, null)
        if (path.isNullOrEmpty()) {
            val file = getBackupFile(context, bookUID, exportParams)
            return file.toUri()
        }
        return path.toUri()
    }

    fun getBackupList(context: Context, bookUID: String): List<File> {
        val backupFiles = getBackupFolder(context, bookUID).listFiles() ?: return emptyList()
        backupFiles.sortDescending()
        return backupFiles.toList()
    }

    fun schedulePeriodicBackups(context: Context) {
        Timber.i("Scheduling backups")
        val request: WorkRequest =
            PeriodicWorkRequest.Builder(BackupWorker::class.java, 1, TimeUnit.DAYS)
                .setInitialDelay(15, TimeUnit.MINUTES)
                .build()

        WorkManager.getInstance(context)
            .enqueue(request)
    }

    fun backupBookAsync(activity: Activity?, bookUID: String, after: BooleanCallback) {
        object : AsyncTask<Any, Any, Boolean>() {
            private var progressDialog: ProgressDialog? = null

            override fun onPreExecute() {
                if (activity != null) {
                    progressDialog = GnucashProgressDialog(activity).apply {
                        setTitle(R.string.title_create_backup_pref)
                        setCancelable(true)
                        setOnCancelListener(DialogInterface.OnCancelListener { dialogInterface: DialogInterface ->
                            cancel(true)
                        })
                        show()
                    }
                }
            }

            override fun doInBackground(vararg objects: Any?): Boolean {
                return backupBook(activity!!, bookUID)
            }

            override fun onPostExecute(result: Boolean) {
                try {
                    if (progressDialog != null && progressDialog!!.isShowing) {
                        progressDialog!!.dismiss()
                    }
                } catch (_: IllegalArgumentException) {
                    //TODO: This is a hack to catch "View not attached to window" exceptions
                    //FIXME by moving the creation and display of the progress dialog to the Fragment
                } finally {
                    progressDialog = null
                }
                if (!result) {
                    activity?.snackLong(R.string.toast_backup_failed)
                }
                after.invoke(result)
            }

            override fun onCancelled() {
                after.invoke(false)
            }
        }.execute()
    }

    fun backupActiveBookAsync(activity: Activity?, after: BooleanCallback) {
        backupBookAsync(activity, GnuCashApplication.activeBookUID!!, after)
    }
}
